### ISSUE TITLE
Feature: Recommend ocamlc as driver for compiling C sources for stublibs

### ISSUE THREAD

**Issue by vicuna** [ocaml/ocaml#5733]
**Original bug ID:** 5733
**Reporter:** gerd
**Status:** confirmed (set by @damiendoligez on 2012-11-15T14:30:52Z)
**Resolution:** open
**Priority:** normal
**Severity:** feature
**Version:** 4.00.0
**Category:** documentation
**Tags:** junior_job
**Monitored by:** @ygrek @hcarty @dbuenzli

## Bug description

Basically, it is up to the user how to call the C compiler for compiling stublibs. However, in many build systems I see that users do it the wrong way (e.g. hardcode the compiler "gcc", or forget to pass switches like -fPIC which are sometimes required). My wish is that the section "Interfacing C with OCaml" includes examples how to do it right - which is absolutely simple, because you just need to use ocamlc as driver even for C sources.

There could e.g. be a new section before "Statically linking C code with OCaml code" about "Compiling C code". Contents:

 - How to use ocamlc as driver for C sources: Sample command
 - Mention that the stdlib is automatically in the include path
 - Mention that the switches are ok for dynamically loadable code
 - The created .o files can be args of ocamlmklib
 - use -I, -ccopt, -cclib to pass args
 - Mention that ocamlc -config can be used to extract the compiler path and switches if lower-level calls are needed

**Comment by gasche** [comment]
This is an interesting improvement proposal for the documentation, and also a good task for a new contributor. Reopening.

**Comment by moyodiallo** [comment]
Hi @gasche, I'm new and I want to work on this.

**Comment by gasche** [comment]
Please feel free to work on this. Note that writing a good proposal for the manual probably requires some familiarity with the C/OCaml FFI, so this may be one of the harder "good first issue" tasks.

**Comment by moyodiallo** [comment]
Cool. I have some knowledge about C/OCaml FFI.

**Comment by tusharhqq** [comment]
i think this issue is solved

[OCaml - Interfacing C with OCaml](https://ocaml.org/manual/intfc.html)

**Comment by adithn875** [comment]
I checked the current manual (5.4) and the specific content described in this issue (using ocamlc as a C compiler driver, -ccopt/-cclib flags, ocamlc -config) does not appear to be documented yet, so this remains open. I'd like to revisit this once I have more FFI experience.

**Comment by OlivierNicole** [comment]
There is a paragraph:

>  The file curses_stubs.c can be compiled with:
> 
>         cc -c -I`ocamlc -where` curses_stubs.c
> 
> or, even simpler,
> 
>         ocamlc -c curses_stubs.c
> 
> (When passed a .c file, the ocamlc command simply calls the C compiler on that file, with the right -I option.)

But I agree that the recommendation should be stronger, and appear sooner in this manual page.

**Comment by moyodiallo** [comment]
> But I agree that the recommendation should be stronger, and appear sooner in this manual page.

Yes agree, one of the argument we could put forward is -fPIC vs -fno-PIC, ocaml compiler have -fPIC as default which is not the case with gcc. And using the ocaml compiler to compile those bindings bring stability since the link flags are harmonized.
